Search indexing (Part 1) - Docs restructure

fonty1 opened the pull request:
**Part 1 - Restructure**
It's long but it's almost entirely text replacements. A quick sanity check is fine.

A Docs exit (Dexit) of the docs site map object (formerly `navLinks`) now called `docsMap` (`$docs/docsMap.js`). `docsMap` is what the indexer, docs navigation and page turner components digest to do their job. I also extracted the raw docs links from the routes definition file and put them in `docsLinks` (`$shared/../links`). These were necessary steps due to the needs of the indexing script.

The structure of the `docsMap` includes some 'convenience properties' for the indexer to chew on which are important for generating the all the information required for each search result (https://share.goabstract.com/53140352-e4a1-4755-a811-84d464cc6d50?mode=design&sha=cf9b31f3cb25734146d94b7f47ce8f0807427d82). 

I also added:
- API explorer to be a top level section
- SDK split into 3 separate pages (as per design)
- DU docs flag from `DATA_UNIONS` to `DATA_UNIONS_DOCS`

**Part 2 - Indexer**
- The actual node indexer script.
- Handle docs page link/duplication

**Part 3 - CI & loose ends**
- Integrating the node script into CI
- Tests for robustness
- Add guidelines/docs for how to add to the docs and keep the indexer happy

juhah commented:
There are some case-sensitive paths:
<img width="542" alt="Screen Shot 2020-04-30 at 10 30 04" src="https://user-images.githubusercontent.com/1064982/80684009-b81c1b80-8acd-11ea-9ffc-4f78e60ff2c8.png">
<img width="1132" alt="Screen Shot 2020-04-30 at 10 24 36" src="https://user-images.githubusercontent.com/1064982/80684029-be11fc80-8acd-11ea-8566-62735f0875ee.png">

This works (`Sdks` -> `SDKs`):
<img width="559" alt="Screen Shot 2020-04-30 at 10 30 20" src="https://user-images.githubusercontent.com/1064982/80684063-c833fb00-8acd-11ea-83e2-6a2adfa90881.png">

juhah commented on app/src/app/Docs.jsx:
```diff
@@ -3,210 +3,203 @@
+    <Route exact path={links.technicalNotes} component={TechnicalNotes} key="technicalNotes" />,
     // Docs Root
-    <Redirect exact from={docs.main} to={docs.introduction} key="DocsRoot" />,
+    <Redirect exact from="/docs" to={links.introduction} key="DocsRoot" />,
```
why hardcoded url here?

fonty1 replied:
Must have been a copy/paste error, fixed now.

fonty1 commented:
> There are some case-sensitive paths:


Ah that's why its working locally and failing here! Will address. Thanks

juhah commented on app/test/unit/app/Docs.test.jsx:
```diff
@@ -216,10 +216,37 @@ jest.mock('$docs/components/DocsPages/Marketplace/IntroToMarketplace', () => ({
     ),
 }))
-jest.mock('$docs/components/DocsPages/SDKs', () => ({
+jest.mock('$docs/components/DocsPages/Sdks/SdkOverview', () => ({
```
Casing here as well:
<img width="781" alt="Screen Shot 2020-04-30 at 10 45 19" src="https://user-images.githubusercontent.com/1064982/80685352-c3704680-8acf-11ea-8d7e-1d086da10a21.png">

fonty1 commented:
@juhah please take another look. Hopefully Git saw all those case changes. https://github.com/streamr-dev/streamr-platform/pull/983/commits/d11bb645db3f3d5a9fba83c90ed3ffe20e92c298

I did a general consistency check to avoid these sorts of bugs in the future - 
SDK -> Sdk
sdks -> sdk
API -> Api

fonty1 commented:
I'm having issues switching between branches, so it seems git can't handle pure casing changes that well. I will fix that now.

juhah commented:
@fonty1 Was this the error that you were referring to:

<img width="700" alt="Screen Shot 2020-05-04 at 11 45 34" src="https://user-images.githubusercontent.com/1064982/80949490-cdaa8180-8dfc-11ea-82d3-361e233230e3.png">

<img width="202" alt="Screen Shot 2020-05-04 at 11 45 38" src="https://user-images.githubusercontent.com/1064982/80949506-d438f900-8dfc-11ea-803b-41218e107b9c.png">

The folder names are `DataUnionsSDK` and `DataUnionsCore` (missing `In`).

fonty1 commented:
> @fonty1 Was this the error that you were referring to:
> The folder names are `DataUnionsSDK` and `DataUnionsCore` (missing `In`).

Sort of :) At least for me, filename casing changes (SDK -> sdk) do not get recognised on their own, so I needed to change their names to get the casing update to register otherwise we run into problems.. then I missed some references inbetween. 

I changed `DocsPages` to `Pages` and removed the `in` from content paths in a roundabout way to get past this.

fonty1 commented:
@juhah all clear now

mondoreale commented on app/src/docs/content/gettingStarted/getBuilding.mdx:
```diff
@@ -1,24 +1,24 @@
 ### Tutorials
-- <Link to={links.docs.tutorials.buildingPubSub}>Build a simple pub/sub system</Link>
-- <Link to={links.docs.tutorials.buildingCustomModule}>Build a custom canvas module</Link>
+- <Link to={docsLinks.buildingPubSub}>Build a simple pub/sub system</Link>
+- <Link to={docsLinks.buildingCustomModule}>Build a custom canvas module</Link>
```
Updated required:

`docsLinks.buildingPubSub` → `docsLinks.buildingPubSubTutorial`
`buildingCustomModule ` → `docsLinks.customModuleTutorial`

mondoreale commented:
I also hear @juhah. I'd love to see `routes` in current or improved version being used across the app. I find it quite handy. [`PLATFORM-1360`](https://streamr.atlassian.net/browse/PLATFORM-1360) is all about the migration.

fonty1 commented:
Thanks @juhah @mondoreale those bugs are fixed. 

I tried out the indexer again and realised I had to separate the `docsLinks` even further, now into its own file in https://github.com/streamr-dev/streamr-platform/pull/983/commits/93aa9211924f07d51bdc72428050f8cf69d072b8. Maybe that makes ticket [PLATFORM-1360](https://streamr.atlassian.net/browse/PLATFORM-1360) easier I hope!

I'll merge this now, ping me if anything doesn't fit right.
